### PR TITLE
ABCI Query

### PR DESCRIPTION
**Pull request by ruseinov**
Work on https://github.com/iov-one/weave/issues/375

### CODE REVIEW

File: client/query_impl.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"fmt"
+
+	"github.com/iov-one/tendermint/rpc/client"
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/errors"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+func NewRPCABCIQuery(c client.Client) Query {
```
Comment (alpe) [inline client/query_impl.go]: pp: `NewABCIQuery` would be enough.

File: client/query_impl.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+// RPCABCIQuery exposes query interface via tendermint rpc
+type RPCABCIQuery struct {
+	conn client.Client
+}
+
+func (q *RPCABCIQuery) Get(key []byte) []byte {
+	query, err := q.conn.ABCIQuery("/", key)
+	if err != nil {
+		panic(err)
```
Comment (alpe) [inline client/query_impl.go]: req: we should not make assumptions how this client package is used. Returning the error is the best practice in Go unless you can provide it some other way.
Please revisit all panics!

File: client/query_impl.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+func (q *RPCABCIQuery) ReverseIterator(start, end []byte) weave.Iterator {
+	panic("implement me")
+}
+
+func NewAppABCIQuery(app abci.Application) Query {
+	return &AppABCIQuery{app}
+}
+
+// AppABCIQuery exposes query interface via abci app query
+type AppABCIQuery struct {
```
Comment (alpe) [inline client/query_impl.go]: Q: And wondering, was there any discussion to write a client that interacts with ABCI apps directly? Which means effectively bypassing Tendermint? The only use case for that I can think of would be tests only. Given the amount of work with the parent ticket I would also depriotize this until the tendermint based client is completed.

File: client/query_impl.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/iov-one/tendermint/rpc/client"
```
Comment (alpe) [inline client/query_impl.go]: req: please use `github.com/tendermint/tendermint/rpc/client` instead.

File: client/query_impl.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	return r
+}
+
+func (q *RPCABCIQuery) Iterator(start, end []byte) weave.Iterator {
+	panic("implement me")
+}
+
+func (q *RPCABCIQuery) ReverseIterator(start, end []byte) weave.Iterator {
+	panic("implement me")
+}
```
Comment (alpe) [inline client/query_impl.go]: req: `func (q *RPCABCIQuery) Has(key []byte) bool ` is missing to satisfy the `weave.ReadOnlyKVStore` interface

**Comment by ethanfrey** [comment]
Can we close this?

I covered this feature in [2586e88](https://github.com/iov-one/weave/pull/549/commits/2586e88d95847e5a2747150c5d10772c5afe9101)

**Comment by ruseinov** [comment]
closing this now

**Pull request closed without merging**